packages/common/src/utils/resolveDynamicValues.ts

```diff
@@ -0,0 +1,182 @@
+  context: IArcGISContext
+): Promise<DynamicValues> {
+  const result = {};
+  // TODO: This could use the context.portal values to skip even a single request
+  // get the portalSelf via memoized function
+  const portal = await memoizedPortalSelf(
+    context.requestOptions.authentication
+  );
+
+  // get the value from the portal and attach into the result
+  const value = getProp(portal, valueDef.sourcePath);
+  setProp(valueDef.outPath, value, result);
+
+  return result;
+}
+
+/**
+ * Resolve a set of dynamic values based on an item query.
+ * This is used to query for a set of items, extract a specific value from each
+ * then run an aggregation on the set of values.
+ * @param valueDef
+ * @param context
+ * @returns
+ */
+async function resolveItemQueryValues(
+  valueDef: IDynamicItemQueryDefinition,
+  context: IArcGISContext
+): Promise<DynamicValues> {
+  const result = {};
+  // const start = Date.now();
+  // execute the query
+  if (valueDef.options) {
+    const opts: ISearchOptions = {
+      ...(valueDef.options as ISearchOptions),
+      authentication: context.requestOptions.authentication,
+    };
+    const response = await memoizedItemSearch(opts);
+    const promises = await response.results.reduce(
+      async (valsPromise: any, r) => {
+        const vals = await valsPromise;
+        const v = getProp(r, valueDef.sourcePath);
+        if (typeof v === "string" || typeof v === "number") {
+          vals.push(Promise.resolve(v));
+        } else {
+          // v is itself a valueDefinition so we call resolveDynamicValues again
+          const vResult = await resolveDynamicValues([v], context);
+
+          // vals needs just the value - the number or string but vResult is an object
+          // so we need to get it out of the object
+          vals.push(getProp(vResult, v.outPath));
+        }
+
+        return vals;
+      },
+      Promise.resolve([] as any[])
+    );
+    const values = await Promise.all(promises);
+    const aggretate = aggregateValues(values, valueDef.aggregation);
+    setProp(valueDef.outPath, aggretate, result);
+  }
+  // timings.staticValues = timings.staticValues + (Date.now() - start);
+  return result;
+}
+
+/**
+ * Resolve a dynamic value that is derived from a service query
+ * @param valueDef
+ * @param context
+ * @returns
+ */
+async function resolveServiceQueryValues(
+  valueDef: IDynamicServiceQueryDefinition,
+  context: IArcGISContext
+): Promise<DynamicValues> {
+  // If no where is provided, default to "1=1"
+  valueDef.options.where = valueDef.options.where || "1=1";
+  // construct the result object
+  const result: DynamicValues = {};
+
+  // Construct the stats definition
+  const statsDef: IStatisticDefinition = {
+    onStatisticField: valueDef.options.field,
+    statisticType: valueDef.options.statisticType,
+    outStatisticFieldName: valueDef.options.field,
+  };
+  // Execute the query
+  const response = await queryFeatures({
```

Comment:
per other comment if we need error handling. The layer could be not available (unshared or just down) or the query not valid due to changed schema, outStats not working (this happens with DC all the time)

Reply:
100% agreed - this is a draft PR just to get input on...
- does this make sense? 
- can it handle the various scenarios we thing it should? 
- can it be extended to work with a metrics api down the line 
- etc etc.

Reply:
ok - That's fine, and as noted in Guide we can have some simple, clear status / next action statements in the guide and likely the files too.